src/compute/src/sink/persist_sink.rs

```diff
@@ -562,6 +567,112 @@ impl<T, R> CorrectionBuffer<T, R> {
+        updates.sort_unstable_by_key(|(_, time, _)| *time);
+
+        let mut updates = updates.into_iter().peekable();
+        while let Some(&(_, time, _)) = updates.peek() {
+            let data = updates
+                .peeking_take_while(|(_, t, _)| *t == time)
+                .map(|(d, _, r)| (d, r));
+
+            use std::collections::btree_map::Entry;
+            match self.0.entry(time) {
+                Entry::Vacant(entry) => {
+                    entry.insert(data.collect());
+                }
+                Entry::Occupied(mut entry) => {
+                    entry.get_mut().extend(data);
+                }
+            }
+        }
+    }
+
+    /// Remove all updates before the given frontier and return an iterator over them.
+    fn drain_before(
+        &mut self,
+        frontier: &Antichain<Timestamp>,
+    ) -> impl Iterator<Item = (D, Timestamp, Diff)> {
+        let retain = match frontier.as_option() {
+            Some(time) => self.0.split_off(time),
+            None => BTreeMap::new(),
+        };
+        let drain = std::mem::replace(&mut self.0, retain);
+        drain
+            .into_iter()
+            .flat_map(|(t, data)| data.into_iter().map(move |(d, r)| (d, t, r)))
+    }
+}
+
+/// A vector that consolidates its contents.
+///
+/// The vector is filled with updates until it reaches capacity. At this point, the updates are
+/// consolidated to free up space. This process repeats until the consolidation recovered less than
+/// half of the vector's capacity, at which point the capacity is doubled.
+struct ConsolidatingVec<D> {
+    inner: Vec<(D, Diff)>,
+}
+
+impl<D: Ord> ConsolidatingVec<D> {
+    fn push(&mut self, item: (D, Diff)) {
+        let capacity = self.inner.capacity();
+        if self.inner.len() == capacity {
+            // The vector is full. First, consolidate to try to recover some space.
+            consolidate(&mut self.inner);
+
+            // If consolidation didn't free at least half the available capacity, double the
+            // capacity. This ensures we won't consolidate over and over again with small gains.
+            if self.inner.len() > capacity / 2 {
+                self.inner.reserve(capacity);
+            }
+        }
+
+        self.inner.push(item);
+    }
+}
+
+impl<D> IntoIterator for ConsolidatingVec<D> {
+    type Item = (D, Diff);
+    type IntoIter = std::vec::IntoIter<(D, Diff)>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+impl<D> FromIterator<(D, Diff)> for ConsolidatingVec<D> {
+    fn from_iter<I>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = (D, Diff)>,
+    {
+        Self {
+            inner: Vec::from_iter(iter),
+        }
+    }
+}
+
+impl<D: Ord> Extend<(D, Diff)> for ConsolidatingVec<D> {
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = (D, Diff)>,
+    {
+        for item in iter {
+            self.push(item);
+        }
```

Comment:
This has the potential deficit that it costs `self.push` for every item, unless we can pre-size the inner vector. Mitigated by `push` being $O(1)$ amortized?

Reply:
We can maybe do better by figuring out the sizes of batches that we push here. I didn't try to go there because (a) as you say, `push` will be O(1) usually and (b) `ConsolidateBuffer` does the same thing and seems fine with it. I think working with batches of known sizes makes the code more complicated (both the `ConsolidatingVec::push` and `UpdateStash::insert`) so I'd like to leave it as is unless we notice performance issues.